cmd/snap-confine/mount-support.c

```diff
@@ -524,29 +524,26 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 		sc_do_mount(SNAP_MOUNT_DIR, dst, NULL, MS_BIND | MS_REC, NULL);
 		sc_do_mount("none", dst, NULL, MS_REC | MS_SLAVE, NULL);
 	}
-	// Create the hostfs directory if one is missing. This directory is a part
-	// of packaging now so perhaps this code can be removed later.
-	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
-	if (mkdir(SC_HOSTFS_DIR, 0755) < 0) {
-		if (errno != EEXIST) {
-			die("cannot perform operation: mkdir %s",
-			    SC_HOSTFS_DIR);
-		}
-	}
-	(void)sc_set_effective_identity(old);
-	// Ensure that hostfs isgroup owned by root. We may have (now or earlier)
-	// created the directory as the user who first ran a snap on a given
-	// system and the group identity of that user is visilbe on disk.
+	// Ensure that hostfs exists and is group-owned by root. We may have (now
+	// or earlier) created the directory as the user who first ran a snap on a
+	// given system and the group identity of that user is visible on disk.
 	// This was LP:#1665004
 	struct stat sb;
 	if (stat(SC_HOSTFS_DIR, &sb) < 0) {
-		die("cannot stat %s", SC_HOSTFS_DIR);
-	}
-	if (sb.st_uid != 0 || sb.st_gid != 0) {
-		if (chown(SC_HOSTFS_DIR, 0, 0) < 0) {
-			die("cannot change user/group owner of %s to root",
-			    SC_HOSTFS_DIR);
+		if (errno == ENOENT) {
+			// Create the hostfs directory if one is missing. This directory is a part
+			// of packaging now so perhaps this code can be removed later.
+			if (mkdir(SC_HOSTFS_DIR, 0755) < 0) {
+				die("cannot perform operation: mkdir %s", SC_HOSTFS_DIR);
+			}
+			if (chown(SC_HOSTFS_DIR, 0, 0) < 0) {
```

Comment:
Are we sure this `chown(,0,0)` will always succeed? I notice you have dropped the code which restored the root group identity above so is it possible this call may now fail without this permission?

Reply:
The call will fail unless we are root or have CAP_CHOWN. In this case we are running under the root user (effectively), but our effective group is the user's: that's why the old code was calling
```
sc_set_effective_identity(sc_root_group_identity())
```
So, the old code was changing the **group** to be 0, in order to not have to worry about permissions (though it was still calling chown, but that was a historical leftover). Now, the plan is to migrate snap-confine to not be running as root and use capabilities instead, so this change aligns towards that goal.

For now, the chmod() call succeeds because we are seteuid 0; in the future, it will succeed because we'll have CAP_CHOWN.